### ISSUE TITLE
test against simple internet taxsim reforms

### ISSUE THREAD

**Issue by MattHJensen** [open-source-economics/Tax-Calculator#426]
The SIMTAX testing facility allows us to test our result against internet TAXSIM's. We should test both the baseline as well as a few reforms. Good reforms to start with would be these:

```
18	
0	No surtax on top rate
x	Surtax for the top rate (such as .05)
19	
0	No surtax on the 2nd to last bracket
x	Surtax for the rate below the top rate
75	
Y	Add $Y to Standard Deduction (may be negative)
0	No change to Standard deduction.
```

http://users.nber.org/~taxsim/taxsim-calc9/plans.html

@GoFroggyRun, could you take the lead on this? 

cc @martinholmer, @feenberg

**Comment by GoFroggyRun** [comment]
Sure. Got you.

**Comment by MattHJensen** [comment]
Testing itemized deductions by discontinuing the phaseout and testing AMT by eliminating it would also be useful. 
```
3	
0	Itemized deductions phaseout continued after 2010
1	Itemized deductions phaseout discontinued after 2010
50	
0	Suppress AMT
1	Include AMT
```

**Comment by martinholmer** [comment]
@MattHJensen and Sean (@GoFroggyRun) Wang,
I may be mistaken, but I don't think that TAXSIM can be used to simulate tax policy reforms.  You should check with Dan @feenberg before starting any work on this.  (There is a limited ```option``` code system in Internet TAXSIM, but, as far as I know, there is not a general capability of specifying policy reforms.)

**Comment by MattHJensen** [comment]
@martinholmer, you are right that there is no general capability for specifying policy reforms in internet TAXSIM. The resolution of this issue would rely on using the internet TAXSIM `option` code system, and the numbers in my two comments above refer to the `options` that would need to be used to simulate the reform in question.

**Comment by martinholmer** [comment]
@MattHJensen said:
> you are right that there is no general capability for specifying policy reforms in internet TAXSIM. The resolution of this issue would rely on using the Internet TAXSIM option code system, and the numbers in my two comments above refer to the options that would need to be used to simulate the reform in question.

OK, then that will work.

**Comment by feenberg** [comment]
On Thu, 29 Oct 2015, Martin Holmer wrote:

> 
> @MattHJensen and Sean (@GoFroggyRun) Wang,
> I may be mistaken, but I don't think that TAXSIM can be used to simulate tax policy
> reforms. You should check with Dan @feenberg before starting any work on this. (There
> is a limited option code system in Internet TAXSIM, but, as far as I know, there is
> not a general capability of specifying policy reforms.)

That is correct about internet taxsim. For reforms there are the options, 
or writing a fortran driver for the calculator and modifying the 
calculator.

dan

**Comment by GoFroggyRun** [comment]
It seems that abandoning the option`56 1` in TAXSIM does have some impacts on EITC, which makes it hard to verify the reforms mentioned.
@martinholmer, I'm wondering is it possible to ignore those observations who have different EITC when counting the differences? Please advice.

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun) said:
> It seems that abandoning the option ```56 1``` in TAXSIM does have some impacts on EITC, which makes it hard to verify the reforms mentioned.

Yes, that is why the validation README.md instructions say to use option ```56 1``` when using TAXSIM.

> @martinholmer, I'm wondering is it possible to ignore those observations who have different EITC when counting the differences?  Please advice.

Why not just used TAXSIM option ```56 1``` and eliminate all that extra work?

**Comment by GoFroggyRun** [comment]
@martinholmer Thanks for your advice! This is really helpful. I didn't realize that at all beforehand.

**Comment by GoFroggyRun** [comment]
Thanks to @martinholmer 's suggestion, the tests against simple reforms become a lot easier. 

So the reforms that have been tested are 
<ol>
  <li>5% surtax on the top rate using a13 and b14 samples</li>
  <li>7% surtax on the top rate using c13 sample</li>
  <li>3% surtax on the second top rate using a14 and b13 samples</li>
  <li>5% surtax on the second top rate using c14 sample</li>
  <li>100 being added to standard deduction using a13 sample</li>
  <li>100 being deducted from standard deduction using b14 sample</li>
  <li>500 being added to standard deduction using c13 sample</li>
  <li>AMT being suppressed using a14, b13 and c13 samples</li>
</ol>
All the results come clean. 

One thing to notice is that, for the AMT option, the correct boolean option to be used to suppress the AMT should be `1`, instead of `0`. Also, the switch option `3` regarding itemized deduction phaseout seems to be null in the way that comparing respective outcomes of a certain file using option `3 1` and `3 0` yield no difference at all.

cc @MattHJensen

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun),  I'm glad this has worked out so well.

Now, it seems to me, we need to think a bit about how to document what you have done, and possibly, put these reform tests into their own script so they can be easily run again and again as Tax-Calculator code is developed.  Such repeated running would serve as a test against future code development causing unintended regressions in the capability of the Tax-Calculator.

Can you describe just briefly what your work involved?  For example, did you use the SIMTAX ```--reform``` option to generate SIMTAX output to compare with TAXSIM output?  If so, could you post the text of those reform files in this Issue discussion?  I think it would be good for everybody on the OSPC team to see  how to simulate policy reforms.  Next to the text of each reform file, please provide the TAXSIM options you used to generate output (noting the corrections you mention above), the input sample(s) used, and the SIMTAX command-line options used (which, I assume, was just ```--taxsim2441```).  Also, provide for each reform the ovar[4] taxdiffs line(s) that show how the SIMTAX income tax liabilities differ from those generated by Internet TAXSIM.

Then, next week will talk with others on the OSPC team about how to use the information that you will have posted on this Issue.   Thanks again for the work on this important issue.

**Comment by GoFroggyRun** [comment]
@martinholmer , unfortunately, I didn't use the `--reform` option in SIMTAX and, instead, made all the reforms manually within the code lines. The SIMTAX option being used is indeed `--taxsim2441` for all tests, and I'll omit them in the results below. 

|Policy Reforms | Python Code Changes | TAXSIM Option | Difference output for ovar[4]|
|:---------------------:|:--------------------------------:|:----------------------:|----------------------|
|5% surtax on 7th bracket| [`II_rt7`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L554) to 0.446 | `56 1 18 0.05` |`A13:280    280     -0.01 [1591]`<br />`B14:679    679  -0.01 [3592]`|
|7% surtax on 7th bracket| [`II_rt7`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L554) to 0.466 | `56 1 18 0.07` |`C13:676    676     -0.01 [258]`|
|3% surtax on 6th bracket| [`II_rt6`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L540) to 0.38 | `56 1 19 0.03` |`A14:236    236    -0.01 [3585]`<br />`B13:819    819    -0.01 [935]`|
|5% surtax on 6th bracket| [`II_rt6`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L540) to 0.40 | `56 1 19 0.05` |`C13:679    679     -0.01 [733]`|
|100 extra on std deduction| [`STD`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L630) add **100**| `56 1 75 100`|`A13:281    281     -0.01 [1591]`|
|100 less on std deduction| [`STD`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L630) minus **100**| `56 1 75 -100`|`B14:717    717     -0.01 [3592]`|
|500 extra on std deduction| [`STD`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L630) add **500**| `56 1 75 500`|`C13:715    715     -0.01 [737]`|
|AMT being suppressed| [`c63200`](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/functions.py#L1029-L1031) to zero| `56 1 50 1`|`A14:303    303     -0.01 [469]`<br />`B13:645    645     -0.01 [687]`<br />`C13:609    609     -0.01 [533]`|

@MattHJensen @martinholmer please take a look and let me know what else might be helpful.

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun) said:
> unfortunately, I didn't use the ```--reform``` option in SIMTAX and, instead, made all the reforms manually within the code lines. 

That is an OK approach for the first time, but this approach is not easily reproducible, which is a very big problem going forward.  Can you try using ```--reform``` files to specify each reform to confirm that you can get the same results as you report above?  (If you have any questions about how to write reform files, I'd be happy to answer them.)

We eventually want to be in a position where we can automate the reform tests you did originally by hand.  Having these reform tests automated will provide additional checks on future developments to Tax-Calculator logic.  And, in addition, using ```--reform``` files will provide some much needed testing of the Policy class ```implement_reform``` method.

All this may seem like unnecessary work to you, but I've had messy, and hugely time-consuming, experiences in the past that were caused by the fact that I didn't have such an automated reform validation capability.

cc @MattHJensen

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun),  Perhaps we should discuss some of the options that would arise in automating the reform tests.  If we do this now, it may eliminate some false steps in the process of automating the reform tests.  Do you want to talk on the phone for a few minutes?  If so, let me know and we will set up a time for me to call you.

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun),
How does something like the following look for the ```validation/reform_tests``` script?
(Notice that is assumes a file naming scheme for the reform files and the output files.)
```
#!/bin/bash
# Executes REFORM validation TESTS using OSPC SimpleTaxIO class as follows:
# (1) generating a random sample of tax filing units (INPUT),
# (2) generating OUTPUT from INPUT using simtax.py, and
# (3) generating tax-difference tabulations by comparing OUTPUT with
#     output file generated by Internet TAXSIM using the same INPUT and the
#     same reform.
# .... check number of command-line arguments
if [[ "$#" -gt 0 ]]; then
    echo "ERROR: no command-line arguments allowed"
    exit 1
fi
# .... specify simtax.py options appropriate for validation tests
OPTIONS="--taxsim2441"
# .... execute reform validation tests
clear
echo "REFORM validation TESTS using OSPC Tax-Calculator SimpleTaxIO class"
echo "(the tests compare OSPC and Internet TAXSIM results)"
for LYY in c13
do
    echo "`date` : $LYY"
    L=${LYY:0:1}    
    YY=${LYY:1:2}
    tclsh make-in.tcl 20$YY $L > $LYY.in
    for REF in 18_0.05 18_0.07 19_0.03 19_0.05 75_100 75_-100 75_500 50_1
    do
        echo "`date` : reform $REF"
        python ../../simtax.py $OPTIONS --reform reform_$REF.json $LYY.in
        mv $LYY.in.out-simtax $LYY.in.out-simtax-$REF
        unzip -oq out-taxsim-reform.zip $LYY.in.out-taxsim-$REF
        tclsh taxdiffs.tcl $LYY.in.out-simtax-$REF \
            $LYY.in.out-taxsim-$REF > $LYY.taxdiffs-$REF
        git diff --name-status $LYY.taxdiffs-$REF
        rm $LYY.in.out-simtax-$REF $LYY.in.out-taxsim-$REF
    done
    rm $LYY.in
    echo "`date`"
done
exit 0
```

**Comment by GoFroggyRun** [comment]
@martinholmer sorry for the delay. 
The script looks good to me and the scheme was exactly what I was using. We can try this out after I finished translating these reforms option. 

One thing @MattHJensen and I were discussing today was, since Internet TAXSIM can incorporate more than one reform option, should we test all the reforms using one file (in C group as we discussed) to do the validation? Doing this might save us some time when it comes to multiple reforms, but might blur the bugs as well. So we come out with an idea that breaks the validation into basically two parts: 
first part of the validation should be based on the results that incorporate all the reforms; second stage, which should validate all reforms separately as the script you made, needs to be performed only when stage one reports error/difference. In this way, we are sort of having a balance between time efficiency and power of debugging. 

I think the outputs shouldn't be too hard to get if everything goes well using the `--reform` option. And it's reasonable to assume that, after continuos improvements, our calculator should have minor bugs where the second stage is needed, and thus improves the efficiency in this way. What do you think of this idea?

**Comment by martinholmer** [comment]
Sean (@GoFroggyRun) Wang said:
> One thing @MattHJensen and I were discussing today was, since Internet TAXSIM can incorporate more than one reform option, should we test all the reforms using one file (in C group as we discussed) to do the validation? Doing this might save us some time when it comes to multiple reforms, but might blur the bugs as well. So we come out with an idea that breaks the validation into basically two parts: 
first part of the validation should be based on the results that incorporate all the reforms; second stage, which should validate all reforms separately as the script you made, needs to be performed only when stage one reports error/difference. In this way, we are sort of having a balance between time efficiency and the power of debugging.
>
> I think the outputs shouldn't be too hard to get if everything goes well using the --reform option. And it's reasonable to assume that, after continuos improvements, our calculator should have minor bugs where the second stage is needed, and thus improves the efficiency in this way. What do you think of this idea?

This sounds like an excellent idea.  Let's use the strategy outlined above.

**Comment by GoFroggyRun** [comment]
After #461 being merged this issue will be completely taken care of. @MattHJensen Feel free to close if it also looks good to you. 

cc @martinholmer

**Comment by martinholmer** [comment]
Close issue #426 as this has been resolved by pull requests #460 and #461.
